### ISSUE TITLE
Wrong overload resolution when overload viability depends on generic constraint inferred from callable argument

### ISSUE THREAD

**Issue by sunny-zuo** [astral-sh/ty#4024]
### Summary

ty 0.0.60 incorrectly preserves the return type of a synchronous callable passed through an overloaded ParamSpec decorator. This seems to have been "caused" by [ruff#26761](https://github.com/astral-sh/ruff/pull/26761), which changed this from `Unknown` to the incorrect type. I'm unsure if this should be part of https://github.com/astral-sh/ty/issues/2278 so apologies in advance if this is reported in the wrong location!

### Reproduction ([playground](https://play.ty.dev/8d6580e4-9109-4d5f-84d9-7776bf91afae))

```python
from collections.abc import Awaitable, Callable
from typing import Any, overload


@overload
def make_awaitable[**P, T: Awaitable[Any]](f: Callable[P, T]) -> Callable[P, T]: ...


@overload
def make_awaitable[**P, T](f: Callable[P, T]) -> Callable[P, Awaitable[T]]: ...


def make_awaitable(f: Callable[..., Any]) -> Callable[..., Awaitable[Any]]:
    raise NotImplementedError


@make_awaitable
def f() -> str:
    return ""


async def repro() -> None:
    reveal_type(f())  # str on 0.0.60 / Unknown on 0.0.59 / "typing.Awaitable[str]" from mypy
    await f()  # `str` is not awaitable (invalid-await)
```

### Version

ty 0.0.60

**Comment by carljm** [comment]
Thanks for the report! Definitely a bug.

The problem here is that our callable assignability checks in generic inference still [discard errors due to the hybrid solver setup](https://github.com/astral-sh/ruff/blob/7bdf9a82f3323090f5454705ef66084a3df1bc37/crates/ty_python_semantic/src/types/generics.rs#L3353). In this case that means that rather than discarding the first overload as non-matching, we keep both overloads, then step 5 prefers the first (wrong) overload and discards the second one (which we should have kept).

The scope of the problem here is overload resolution for generic overloads, when overload viability depends on generic constraints inferred through a callable argument.

The right fix here is to address the linked comment and stop discarding unsatisfiable constraints on generic callable assignability. We could probably also do that in a way that is narrowly targeted to the "transparent decorator" case only, in order to address this regression?

Would like some feedback from @dcreager or @dhruvmanila about how this intersects with other ongoing work, before trying to put up a fix.
